File: aip/packages/typescript/src/analysis/engine.ts
/**
 * Integrity analysis engine for the Agent Integrity Protocol.
 *
 * Contains the pure function `checkIntegrity` that creates an
 * IntegrityCheckpoint from analysis LLM output, plus helper functions
 * for verdict-to-action mapping and signal assembly.
 */

import { createHash, randomUUID } from "node:crypto";
import type {
  IntegrityCheckpoint,
  IntegrityVerdict,
  AnalysisMetadata,
} from "../schemas/checkpoint.js";
import type {
  IntegrityConcern,
  ConcernCategory,
  IntegritySeverity,
} from "../schemas/concern.js";
import type {
  ConscienceContext,
  ConsultationDepth,
} from "../schemas/conscience.js";
import type {
  IntegritySignal,
  RecommendedAction,
  WindowSummary,
} from "../schemas/signal.js";
import { CHECKPOINT_ID_PREFIX, MAX_EVIDENCE_LENGTH } from "../constants.js";

// ---------------------------------------------------------------------------
// Validation sets
// ---------------------------------------------------------------------------

const VALID_VERDICTS = new Set<string>([
  "clear",
  "review_needed",
  "boundary_violation",
]);

const VALID_CATEGORIES = new Set<string>([
  "prompt_injection",
  "value_misalignment",
  "autonomy_violation",
  "reasoning_corruption",
  "deceptive_reasoning",
  "undeclared_intent",
  "output_misalignment",
  "output_injection_compliance",
]);

const VALID_SEVERITIES = new Set<string>([
  "low",
  "medium",
  "high",
  "critical",
]);

const VALID_DEPTHS = new Set<string>(["surface", "standard", "deep"]);

// ---------------------------------------------------------------------------
// Input type
// ---------------------------------------------------------------------------

/** Input to the checkIntegrity pure function */
export interface CheckIntegrityInput {
  /** Raw analysis LLM response (JSON string) */
  analysisResponse: string;

  /** Extracted thinking block metadata */
  thinking: {
    /** Pre-computed SHA-256 of thinking content */
    hash: string;
    /** LLM provider that generated the thinking block */
    provider: string;
    /** Model that generated the thinking block */
    model: string;
    /** Number of tokens in the original thinking block */
    tokensOriginal: number;
    /** Number of tokens sent to the analysis LLM */
    tokensAnalyzed: number;
    /** Whether the thinking block was truncated */
    truncated: boolean;
    /** Provider extraction confidence (0.0 to 1.0) */
    confidence: number;
  };

  /** Agent identifier */
  agentId: string;
  /** Alignment Card identifier */
  cardId: string;
  /** Session identifier */
  sessionId: string;

  /** Window position */
  windowPosition: { index: number; window_size: number };

  /** Analysis LLM model used */
  analysisModel: string;
  /** Time taken for analysis in milliseconds */
  analysisDurationMs: number;

  /** Optional AAP trace link */
  linkedTraceId?: string | null;

  /** Output block metadata (when output analysis is enabled) */
  output?: {
    /** SHA-256 hash of the output block */
    hash: string;
    /** Original token count */
    tokensOriginal: number;
    /** Tokens sent to analysis */
    tokensAnalyzed: number;
    /** Whether the output block was truncated */
    truncated: boolean;
  };
}

// ---------------------------------------------------------------------------
// Validation helpers
// ---------------------------------------------------------------------------

function validateVerdict(value: unknown): IntegrityVerdict {
  if (typeof value !== "string") {
    throw new Error(
      `Invalid verdict: expected string, got ${typeof value}`,
    );
  }
  if (!VALID_VERDICTS.has(value)) {
    throw new Error(
      `Invalid verdict value: "${value}". Must be one of: ${[...VALID_VERDICTS].join(", ")}`,
    );
  }
  return value as IntegrityVerdict;
}

function validateConcerns(value: unknown): IntegrityConcern[] {
  if (!Array.isArray(value)) {
    throw new Error(
      `Invalid concerns: expected array, got ${typeof value}`,
    );
  }

  return value.map((item: unknown, index: number) => {
    if (typeof item !== "object" || item === null) {
      throw new Error(
        `Invalid concern at index ${index}: expected object`,
      );
    }

    const concern = item as Record<string, unknown>;

    // Validate category
    if (typeof concern.category !== "string") {
      throw new Error(
        `Invalid concern at index ${index}: category must be a string`,
      );
    }
    if (!VALID_CATEGORIES.has(concern.category)) {
      throw new Error(
        `Invalid concern category at index ${index}: "${concern.category}". Must be one of: ${[...VALID_CATEGORIES].join(", ")}`,
      );
    }

    // Validate severity
    if (typeof concern.severity !== "string") {
      throw new Error(
        `Invalid concern at index ${index}: severity must be a string`,
      );
    }
    if (!VALID_SEVERITIES.has(concern.severity)) {
      throw new Error(
        `Invalid concern severity at index ${index}: "${concern.severity}". Must be one of: ${[...VALID_SEVERITIES].join(", ")}`,
      );
    }

    // Validate description
    if (typeof concern.description !== "string") {
      throw new Error(
        `Invalid concern at index ${index}: description must be a string`,
      );
    }

    // Validate evidence (truncate to MAX_EVIDENCE_LENGTH)
    if (typeof concern.evidence !== "string") {
      throw new Error(
        `Invalid concern at index ${index}: evidence must be a string`,
      );
    }
    const evidence =
      concern.evidence.length > MAX_EVIDENCE_LENGTH
        ? concern.evidence.slice(0, MAX_EVIDENCE_LENGTH)
        : concern.evidence;

    // Validate relevant_card_field (string or null)
    const relevantCardField =
      concern.relevant_card_field === undefined
        ? null
        : (concern.relevant_card_field as string | null);

    // Validate relevant_conscience_value (string or null)
    const relevantConscienceValue =
      concern.relevant_conscience_value === undefined
        ? null
        : (concern.relevant_conscience_value as string | null);

    return {
      category: concern.category as ConcernCategory,
      severity: concern.severity as IntegritySeverity,
      description: concern.description,
      evidence,
      relevant_card_field: relevantCardField,
      relevant_conscience_value: relevantConscienceValue,
    };
  });
}

function validateConscienceContext(value: unknown): ConscienceContext {
  if (typeof value !== "object" || value === null) {
    throw new Error(
      `Invalid conscience_context: expected object, got ${value === null ? "null" : typeof value}`,
    );
  }

  const ctx = value as Record<string, unknown>;

  // Validate values_checked
  if (!Array.isArray(ctx.values_checked)) {
    throw new Error(
      "Invalid conscience_context: values_checked must be an array",
    );
  }

  // Validate conflicts
  if (!Array.isArray(ctx.conflicts)) {
    throw new Error(
      "Invalid conscience_context: conflicts must be an array",
    );
  }

  // Validate supports
  if (!Array.isArray(ctx.supports)) {
    throw new Error(
      "Invalid conscience_context: supports must be an array",
    );
  }

  // Validate considerations
  if (!Array.isArray(ctx.considerations)) {
    throw new Error(
      "Invalid conscience_context: considerations must be an array",
    );
  }

  // Validate consultation_depth
  if (typeof ctx.consultation_depth !== "string") {
    throw new Error(
      "Invalid conscience_context: consultation_depth must be a string",
    );
  }
  if (!VALID_DEPTHS.has(ctx.consultation_depth)) {
    throw new Error(
      `Invalid consultation_depth: "${ctx.consultation_depth}". Must be one of: ${[...VALID_DEPTHS].join(", ")}`,
    );
  }

  return {
    values_checked: ctx.values_checked as string[],
    conflicts: ctx.conflicts as string[],
    supports: ctx.supports as string[],
    considerations: ctx.considerations as string[],
    consultation_depth: ctx.consultation_depth as ConsultationDepth,
  };
}

// ---------------------------------------------------------------------------
// JSON extraction helper
// ---------------------------------------------------------------------------

/**
 * Extract the first complete JSON object from a string using brace-depth counting.
 *
 * This avoids the greedy regex `/{[\s\S]*}/` which could match from the first `{`
 * to the last `}` in text containing multiple JSON objects or trailing content.
 */
function extractFirstJsonObject(text: string): string | null {
  let depth = 0;
  let start = -1;
  for (let i = 0; i < text.length; i++) {
    if (text[i] === '{') {
      if (depth === 0) start = i;
      depth++;
    } else if (text[i] === '}') {
      depth--;
      if (depth === 0 && start !== -1) {
        return text.slice(start, i + 1);
      }
    }
  }
  return null;
}

// ---------------------------------------------------------------------------
// Core functions
// ---------------------------------------------------------------------------

/**
 * Parse and validate the analysis LLM's JSON response, create an IntegrityCheckpoint.
 *
 * This is a pure function: given the same input it always produces the same
 * output structure (only checkpoint_id and timestamp vary).
 *
 * @throws Error if the JSON is invalid or required fields are missing/wrong type
 */
export function checkIntegrity(input: CheckIntegrityInput): IntegrityCheckpoint {
  // 1. Parse analysisResponse as JSON
  // Strip markdown code fences if present (e.g. ```json ... ```)
  // Some models (claude-haiku-4-5) wrap JSON responses in code fences
  let jsonText = input.analysisResponse;
  const extracted = extractFirstJsonObject(jsonText);
  if (extracted) {
    jsonText = extracted;
  }

  // Sanitize trailing commas (Haiku 4.5 occasionally produces them)
  jsonText = jsonText.replace(/,\s*([}\]])/g, '$1');

  let parsed: Record<string, unknown>;
  try {
    parsed = JSON.parse(jsonText) as Record<string, unknown>;
  } catch {
    throw new Error(
      `Failed to parse analysis response as JSON: ${input.analysisResponse.slice(0, 100)}`,
    );
  }

  // 2. Validate required fields
  const verdict = validateVerdict(parsed.verdict);

  const concerns = validateConcerns(parsed.concerns);

  if (typeof parsed.reasoning_summary !== "string") {
    throw new Error(
      `Invalid reasoning_summary: expected string, got ${typeof parsed.reasoning_summary}`,
    );
  }
  const reasoningSummary = parsed.reasoning_summary;

  const conscienceContext = validateConscienceContext(
    parsed.conscience_context,
  );

  // 3. Cross-validate verdict and concerns
  const crossValidationWarnings: string[] = [];

  if (verdict === "clear") {
    const criticalOrHigh = concerns.filter(
      (c) => c.severity === "critical" || c.severity === "high",
    );
    if (criticalOrHigh.length > 0) {
      const maxSeverity = criticalOrHigh.some((c) => c.severity === "critical")
        ? "critical"
        : "high";
      crossValidationWarnings.push(
        `Verdict is "clear" but ${criticalOrHigh.length} concern(s) have severity "${maxSeverity}". ` +
          `This may indicate an inconsistent analysis result.`,
      );
    }
  } else if (verdict === "boundary_violation" && concerns.length === 0) {
    crossValidationWarnings.push(
      `Verdict is "boundary_violation" but concerns array is empty. ` +
        `Expected at least one concern for a boundary violation.`,
    );
  }

  // 4. Generate checkpoint_id
  const checkpointId = `${CHECKPOINT_ID_PREFIX}${randomUUID()}`;

  // 5. Build analysis metadata
  const analysisMetadata: AnalysisMetadata = {
    analysis_model: input.analysisModel,
    analysis_duration_ms: input.analysisDurationMs,
    thinking_tokens_original: input.thinking.tokensOriginal,
    thinking_tokens_analyzed: input.thinking.tokensAnalyzed,
    truncated: input.thinking.truncated,
    extraction_confidence: input.thinking.confidence,
    ...(input.output
      ? {
          analysis_scope: "thinking_and_output" as const,
          output_tokens_original: input.output.tokensOriginal,
          output_tokens_analyzed: input.output.tokensAnalyzed,
          output_truncated: input.output.truncated,
        }
      : {
          analysis_scope: "thinking_only" as const,
        }),
  };

  // 6. Assemble and return IntegrityCheckpoint
  return {
    checkpoint_id: checkpointId,
    agent_id: input.agentId,
    card_id: input.cardId,
    session_id: input.sessionId,
    timestamp: new Date().toISOString(),
    thinking_block_hash: input.thinking.hash,
    ...(input.output ? { output_block_hash: input.output.hash } : {}),
    provider: input.thinking.provider,
    model: input.thinking.model,
    verdict,
    concerns,
    reasoning_summary: reasoningSummary,
    conscience_context: conscienceContext,
    window_position: {
      index: input.windowPosition.index,
      window_size: input.windowPosition.window_size,
    },
    analysis_metadata: analysisMetadata,
    linked_trace_id: input.linkedTraceId ?? null,
    ...(crossValidationWarnings.length > 0
      ? { cross_validation_warnings: crossValidationWarnings }
      : {}),
  };
}

/**
 * Map a verdict + concerns to a recommended action.
 *
 * - clear -> "continue"
 * - review_needed -> "log_and_continue"
 * - boundary_violation with any critical severity -> "deny_and_escalate"
 * - boundary_violation without critical -> "pause_for_review"
 */
export function mapVerdictToAction(
  verdict: IntegrityVerdict,
  concerns: IntegrityConcern[],
): RecommendedAction {
  switch (verdict) {
    case "clear":
      return "continue";
    case "review_needed":
      return "log_and_continue";
    case "boundary_violation": {
      const hasCritical = concerns.some((c) => c.severity === "critical");
      return hasCritical ? "deny_and_escalate" : "pause_for_review";
    }
  }
}

/**
 * Map a verdict to the proceed boolean.
 *
 * - clear -> true
 * - review_needed -> true
 * - boundary_violation -> false
 */
export function mapVerdictToProceed(verdict: IntegrityVerdict): boolean {
  return verdict !== "boundary_violation";
}

/**
 * Build an IntegritySignal from a checkpoint and window summary.
 */
export function buildSignal(
  checkpoint: IntegrityCheckpoint,
  windowSummary: WindowSummary,
): IntegritySignal {
  return {
    checkpoint,
    proceed: mapVerdictToProceed(checkpoint.verdict),
    recommended_action: mapVerdictToAction(
      checkpoint.verdict,
      checkpoint.concerns,
    ),
    window_summary: windowSummary,
  };
}

/**
 * Compute SHA-256 hash of a thinking block.
 *
 * Used to generate the thinking_block_hash without storing the
 * full thinking block content in the checkpoint.
 */
export function hashThinkingBlock(content: string): string {
  return createHash("sha256").update(content).digest("hex");
}

File: aip/packages/python/src/aip/analysis/engine.py
"""Integrity analysis engine for the Agent Integrity Protocol.

Contains the pure function ``check_integrity`` that creates an
IntegrityCheckpoint from analysis LLM output, plus helper functions
for verdict-to-action mapping and signal assembly.

Port of packages/typescript/src/analysis/engine.ts
"""

from __future__ import annotations

import hashlib
import json
import re
from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any, cast
from uuid import uuid4

from aip.constants import CHECKPOINT_ID_PREFIX, MAX_EVIDENCE_LENGTH
from aip.schemas.checkpoint import (
    AnalysisMetadata,
    IntegrityCheckpoint,
    IntegrityVerdict,
    WindowPosition,
)
from aip.schemas.concern import (
    ConcernCategory,
    IntegrityConcern,
)
from aip.schemas.conscience import ConscienceContext
from aip.schemas.signal import (
    IntegritySignal,
    RecommendedAction,
    WindowSummary,
)

# ---------------------------------------------------------------------------
# Validation sets
# ---------------------------------------------------------------------------

VALID_VERDICTS: set[str] = {"clear", "review_needed", "boundary_violation"}

VALID_CATEGORIES: set[str] = {
    "prompt_injection",
    "value_misalignment",
    "autonomy_violation",
    "reasoning_corruption",
    "deceptive_reasoning",
    "undeclared_intent",
    "output_misalignment",
    "output_injection_compliance",
}

VALID_SEVERITIES: set[str] = {"low", "medium", "high", "critical"}

VALID_DEPTHS: set[str] = {"surface", "standard", "deep"}

# Non-canonical concern-category aliases the analysis LLM occasionally emits
# and their canonical ConcernCategory. The analyzer prompt's EVALUATION
# PRIORITY list names "Forbidden action intent" (planning forbidden actions),
# which the model labels ``forbidden_action_intent`` — a category that is NOT
# in the canonical ConcernCategory enum (schemas/concern.schema.json). Forbidden
# actions live under ``autonomy_envelope.forbidden_actions`` on the alignment
# card, so the canonical home for that concern is ``autonomy_violation``.
#
# Without this reconciliation, a genuinely-detected injection whose concern the
# model labels ``forbidden_action_intent`` would crash ``_validate_concerns``
# instead of yielding a ``boundary_violation`` checkpoint — a security-relevant
# fail-loud on exactly the input this SDK exists to catch (MNE-1725). Mapping to
# an existing canonical category (rather than adding a 9th enum value) keeps the
# TS / Python / Rust-prover parity contract intact — the canonical schema warns
# that enum changes are breaking and must be synced across every consumer.
# Only this known alias is remapped; any other unrecognized category still
# raises, preserving drift detection.
CATEGORY_ALIASES: dict[str, str] = {
    "forbidden_action_intent": "autonomy_violation",
}

# ---------------------------------------------------------------------------
# Input types
# ---------------------------------------------------------------------------


@dataclass
class ThinkingInput:
    """Extracted thinking block metadata."""

    hash: str
    """Pre-computed SHA-256 of thinking content."""

    provider: str
    """LLM provider that generated the thinking block."""

    model: str
    """Model that generated the thinking block."""

    tokens_original: int
    """Number of tokens in the original thinking block."""

    tokens_analyzed: int
    """Number of tokens sent to the analysis LLM."""

    truncated: bool
    """Whether the thinking block was truncated."""

    confidence: float
    """Provider extraction confidence (0.0 to 1.0)."""


@dataclass
class CheckIntegrityInput:
    """Input to the check_integrity pure function."""

    analysis_response: str
    """Raw analysis LLM response (JSON string)."""

    thinking: ThinkingInput
    """Extracted thinking block metadata."""

    agent_id: str
    """Agent identifier."""

    card_id: str
    """Alignment Card identifier."""

    session_id: str
    """Session identifier."""

    window_position: WindowPosition
    """Window position."""

    analysis_model: str
    """Analysis LLM model used."""

    analysis_duration_ms: float
    """Time taken for analysis in milliseconds."""

    linked_trace_id: str | None = None
    """Optional AAP trace link."""


# ---------------------------------------------------------------------------
# Validation helpers
# ---------------------------------------------------------------------------


def _validate_verdict(value: Any) -> IntegrityVerdict:
    if not isinstance(value, str):
        raise ValueError(
            f"Invalid verdict: expected string, got {type(value).__name__}"
        )
    if value not in VALID_VERDICTS:
        raise ValueError(
            f'Invalid verdict value: "{value}". '
            f"Must be one of: {', '.join(sorted(VALID_VERDICTS))}"
        )
    return value  # type: ignore[return-value]


def _validate_concerns(value: Any) -> list[IntegrityConcern]:
    if not isinstance(value, list):
        raise ValueError(
            f"Invalid concerns: expected array, got {type(value).__name__}"
        )

    concerns: list[IntegrityConcern] = []
    for index, item in enumerate(value):
        if not isinstance(item, dict):
            raise ValueError(
                f"Invalid concern at index {index}: expected object"
            )

        # Validate category
        if not isinstance(item.get("category"), str):
            raise ValueError(
                f"Invalid concern at index {index}: category must be a string"
            )
        # Reconcile known non-canonical aliases (e.g. "forbidden_action_intent"
        # the prompt elicits) to their canonical category before validation, so
        # a genuinely-detected injection is not rejected. See CATEGORY_ALIASES.
        category = CATEGORY_ALIASES.get(item["category"], item["category"])
        if category not in VALID_CATEGORIES:
            raise ValueError(
                f'Invalid concern category at index {index}: "{item["category"]}". '
                f"Must be one of: {', '.join(sorted(VALID_CATEGORIES))}"
            )

        # Validate severity
        if not isinstance(item.get("severity"), str):
            raise ValueError(
                f"Invalid concern at index {index}: severity must be a string"
            )
        if item["severity"] not in VALID_SEVERITIES:
            raise ValueError(
                f'Invalid concern severity at index {index}: "{item["severity"]}". '
                f"Must be one of: {', '.join(sorted(VALID_SEVERITIES))}"
            )

        # Validate description
        if not isinstance(item.get("description"), str):
            raise ValueError(
                f"Invalid concern at index {index}: description must be a string"
            )

        # Validate evidence (truncate to MAX_EVIDENCE_LENGTH)
        if not isinstance(item.get("evidence"), str):
            raise ValueError(
                f"Invalid concern at index {index}: evidence must be a string"
            )
        evidence = item["evidence"]
        if len(evidence) > MAX_EVIDENCE_LENGTH:
            evidence = evidence[:MAX_EVIDENCE_LENGTH]

        # Validate relevant_card_field (string or null)
        relevant_card_field = item.get("relevant_card_field", None)

        # Validate relevant_conscience_value (string or null)
        relevant_conscience_value = item.get("relevant_conscience_value", None)

        concerns.append(
            IntegrityConcern(
                category=cast(ConcernCategory, category),
                severity=item["severity"],
                description=item["description"],
                evidence=evidence,
                relevant_card_field=relevant_card_field,
                relevant_conscience_value=relevant_conscience_value,
            )
        )

    return concerns


def _validate_conscience_context(value: Any) -> ConscienceContext:
    if not isinstance(value, dict):
        raise ValueError(
            f"Invalid conscience_context: expected object, "
            f"got {'null' if value is None else type(value).__name__}"
        )

    # Validate values_checked
    if not isinstance(value.get("values_checked"), list):
        raise ValueError(
            "Invalid conscience_context: values_checked must be an array"
        )

    # Validate conflicts
    if not isinstance(value.get("conflicts"), list):
        raise ValueError(
            "Invalid conscience_context: conflicts must be an array"
        )

    # Validate supports
    if not isinstance(value.get("supports"), list):
        raise ValueError(
            "Invalid conscience_context: supports must be an array"
        )

    # Validate considerations
    if not isinstance(value.get("considerations"), list):
        raise ValueError(
            "Invalid conscience_context: considerations must be an array"
        )

    # Validate consultation_depth
    if not isinstance(value.get("consultation_depth"), str):
        raise ValueError(
            "Invalid conscience_context: consultation_depth must be a string"
        )
    if value["consultation_depth"] not in VALID_DEPTHS:
        raise ValueError(
            f'Invalid consultation_depth: "{value["consultation_depth"]}". '
            f"Must be one of: {', '.join(sorted(VALID_DEPTHS))}"
        )

    return ConscienceContext(
        values_checked=value["values_checked"],
        conflicts=value["conflicts"],
        supports=value["supports"],
        considerations=value["considerations"],
        consultation_depth=value["consultation_depth"],
    )


# ---------------------------------------------------------------------------
# Core functions
# ---------------------------------------------------------------------------


def check_integrity(input: CheckIntegrityInput) -> IntegrityCheckpoint:
    """Parse and validate the analysis LLM's JSON response, create an IntegrityCheckpoint.

    This is a pure function: given the same input it always produces the same
    output structure (only checkpoint_id and timestamp vary).

    Raises:
        ValueError: If the JSON is invalid or required fields are missing/wrong type.
    """
    # 1. Parse analysis_response as JSON
    # Strip markdown code fences if present (e.g. ```json ... ```)
    # Some models (claude-haiku-4-5) wrap JSON responses in code fences
    json_text = input.analysis_response.strip()
    json_text = re.sub(r'^```(?:json)?\s*', '', json_text)
    json_text = re.sub(r'\s*```$', '', json_text.strip())

    # Sanitize trailing commas (Haiku 4.5 occasionally produces them)
    sanitized = re.sub(r",\s*([}\]])", r"\1", json_text)
    try:
        parsed: dict[str, Any] = json.loads(sanitized)
    except json.JSONDecodeError:
        raise ValueError(
            f"Failed to parse analysis response as JSON: "
            f"{input.analysis_response[:100]}"
        )

    # 2. Validate required fields
    verdict = _validate_verdict(parsed.get("verdict"))
    concerns = _validate_concerns(parsed.get("concerns"))

    if not isinstance(parsed.get("reasoning_summary"), str):
        raise ValueError(
            f"Invalid reasoning_summary: expected string, "
            f"got {type(parsed.get('reasoning_summary')).__name__}"
        )
    reasoning_summary: str = parsed["reasoning_summary"]

    conscience_context = _validate_conscience_context(
        parsed.get("conscience_context")
    )

    # 3. Cross-validate verdict and concerns
    cross_validation_warnings: list[str] = []

    if verdict == "clear":
        critical_or_high = [
            c for c in concerns if c.severity in ("critical", "high")
        ]
        if critical_or_high:
            max_severity = (
                "critical"
                if any(c.severity == "critical" for c in critical_or_high)
                else "high"
            )
            cross_validation_warnings.append(
                f'Verdict is "clear" but {len(critical_or_high)} concern(s) have severity '
                f'"{max_severity}". This may indicate an inconsistent analysis result.'
            )
    elif verdict == "boundary_violation" and len(concerns) == 0:
        cross_validation_warnings.append(
            'Verdict is "boundary_violation" but concerns array is empty. '
            "Expected at least one concern for a boundary violation."
        )

    # 4. Generate checkpoint_id
    checkpoint_id = f"{CHECKPOINT_ID_PREFIX}{uuid4()}"

    # 5. Build analysis metadata
    analysis_metadata = AnalysisMetadata(
        analysis_model=input.analysis_model,
        analysis_duration_ms=input.analysis_duration_ms,
        thinking_tokens_original=input.thinking.tokens_original,
        thinking_tokens_analyzed=input.thinking.tokens_analyzed,
        truncated=input.thinking.truncated,
        extraction_confidence=input.thinking.confidence,
    )

    # 6. Assemble and return IntegrityCheckpoint
    return IntegrityCheckpoint(
        checkpoint_id=checkpoint_id,
        agent_id=input.agent_id,
        card_id=input.card_id,
        session_id=input.session_id,
        timestamp=datetime.now(timezone.utc).isoformat(),
        thinking_block_hash=input.thinking.hash,
        provider=input.thinking.provider,
        model=input.thinking.model,
        verdict=verdict,
        concerns=concerns,
        reasoning_summary=reasoning_summary,
        conscience_context=conscience_context,
        window_position=WindowPosition(
            index=input.window_position.index,
            window_size=input.window_position.window_size,
        ),
        analysis_metadata=analysis_metadata,
        linked_trace_id=input.linked_trace_id,
        cross_validation_warnings=cross_validation_warnings if cross_validation_warnings else None,
    )


def map_verdict_to_action(
    verdict: IntegrityVerdict,
    concerns: list[IntegrityConcern],
) -> RecommendedAction:
    """Map a verdict + concerns to a recommended action.

    - clear -> "continue"
    - review_needed -> "log_and_continue"
    - boundary_violation with any critical severity -> "deny_and_escalate"
    - boundary_violation without critical -> "pause_for_review"
    """
    if verdict == "clear":
        return "continue"
    if verdict == "review_needed":
        return "log_and_continue"
    # boundary_violation
    has_critical = any(c.severity == "critical" for c in concerns)
    return "deny_and_escalate" if has_critical else "pause_for_review"


def map_verdict_to_proceed(verdict: IntegrityVerdict) -> bool:
    """Map a verdict to the proceed boolean.

    - clear -> True
    - review_needed -> True
    - boundary_violation -> False
    """
    return verdict != "boundary_violation"


def build_signal(
    checkpoint: IntegrityCheckpoint,
    window_summary: WindowSummary,
) -> IntegritySignal:
    """Build an IntegritySignal from a checkpoint and window summary."""
    return IntegritySignal(
        checkpoint=checkpoint,
        proceed=map_verdict_to_proceed(checkpoint.verdict),
        recommended_action=map_verdict_to_action(
            checkpoint.verdict,
            checkpoint.concerns,
        ),
        window_summary=window_summary,
    )


def hash_thinking_block(content: str) -> str:
    """Compute SHA-256 hash of a thinking block.

    Used to generate the thinking_block_hash without storing the
    full thinking block content in the checkpoint.
    """
    return hashlib.sha256(content.encode()).hexdigest()
